File: Prototype/cpp-source/PirelliFerrari.cpp
//
// Created by Pavel Akhtyamov on 05/03/2018.
//

#include "PirelliFerrari.h"


PirelliFerrari::PirelliFerrari(const std::string &shassi_name,
                               const std::string &engine_name) :
    FerrariPrototype(engine_name) {
  this->shassi_name = shassi_name;
}

PirelliFerrari::PirelliFerrari(const PirelliFerrari &prototype) :
    FerrariPrototype(prototype) {
  this->shassi_name = prototype.shassi_name;
}

Prototype* PirelliFerrari::clone() const {
  return new PirelliFerrari(*this);
}

File: Prototype/python-source/PirelliFerrari.py
from __future__ import annotations

from FerrariPrototype import FerrariPrototype


class PirelliFerrari(FerrariPrototype):
    """Наследник копируется правильно, потому что копирует себя сам.

    Если бы копию делал вызывающий, он вызвал бы конструктор базового класса и
    молча потерял бы шасси. Здесь потерять его невозможно.
    """

    def __init__(self, chassis_name: str, engine_name: str) -> None:
        super().__init__(engine_name)
        self._chassis_name = chassis_name

    @property
    def chassis_name(self) -> str:
        return self._chassis_name

    def clone(self) -> PirelliFerrari:
        return PirelliFerrari(self._chassis_name, self.engine_name)

    def __str__(self) -> str:
        return f"PirelliFerrari(chassis={self._chassis_name}, engine={self.engine_name})"
